repos/system_upgrade/el7toel8/actors/excluderepositories/libraries/excluderepositories.py

```diff
@@ -0,0 +1,145 @@
+    """
+    opt_repo_mapping = {}
+    repo_map = next(api.consume(RepositoriesMap), None)
+    custom_target_repos = tuple(api.consume(CustomTargetRepository))
+    if repo_map:
+        for repo in repo_map.repositories:
+            if _is_optional_repo(repo) and not _manually_enabled_repo(
+                repo, custom_target_repos
+            ):
+                opt_repo_mapping[repo.from_repoid] = repo.to_repoid
+    return opt_repo_mapping
+
+
+def _get_repos_to_exclude():
+    """Get a list of repoids to not use during the upgrade.
+
+    These are such RHEL 8 repositories that mapped to disabled RHEL7 Optional
+    repositories.
+    """
+    opt_repo_mapping = _get_optional_repo_mapping()
+    repos_to_exclude = []
+    repos_on_system = next(api.consume(RepositoriesFacts), None)
+    for repo_file in repos_on_system.repositories:
+        for repo in repo_file.data:
+            if repo.repoid in opt_repo_mapping and not repo.enabled:
+                repos_to_exclude.append(opt_repo_mapping[repo.repoid])
+    return repos_to_exclude
+
+
+def process():
+    """Exclude the RHEL8 CRB repo if the RHEL7 optional repo is not enabled.
+
+    If manually enabled RHEL8 repo (by --enablerepo option) the repo do not
+    excluded.
+    """
+    repos_to_exclude = _get_repos_to_exclude()
+    if manually_enabled_repos_in_use:
+        _report_using_unsupported_repos(manually_enabled_repos_in_use)
+    if repos_to_exclude:
+        api.current_logger().info(
+            "The optional repository is not enabled. Excluding %r "
+            "from the upgrade",
+            repos_to_exclude,
+        )
+        api.produce(RepositoriesExcluded(repoids=repos_to_exclude))
+
+        report = [
+            reporting.Title("Excluded RHEL 8 repositories"),
+            reporting.Summary(
+                "The following repositories are not supported by "
+                "Red Hat and are excluded from the list of repositories "
+                "used during the upgrade.\n- {}".format(
+                    "\n- ".join(repos_to_exclude)
+                )
+            ),
+            reporting.Severity(reporting.Severity.INFO),
+            reporting.Tags([reporting.Tags.REPOSITORY]),
+            reporting.Flags([reporting.Flags.FAILURE]),
+            reporting.ExternalLink(
+                url=(
+                    "https://access.redhat.com/documentation/en-us/"
+                    "red_hat_enterprise_linux/8/html/package_manifest/"
+                    "codereadylinuxbuilder-repository."
+                ),
+                title="CodeReady Linux Builder repository",
+            ),
+            reporting.Remediation(
+                hint="At your own risk! Use leapp upgrade {}".format(
+                    " ".join(
+                        ["--enablerepo " + repo for repo in repos_to_exclude]
+                    )
+                )
+            ),
+        ]
+        reporting.create_report(report)
```

Comment:
The `process()` function is the highes level function/entry point of the library. Thus, from the clean code perspective, it should not be doing the low level stuff. Add this reporting functionality into a separate function like `report_excluded_repos()` and just call it from `process()`.